// wft-geodb-angular-client/esm5/lib/geodb-pro.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClientModule, HTTP_INTERCEPTORS } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { AuthInterceptor } from './auth.interceptor';
import { GeoDbService } from './geodb.service';
import { GeoDbConfig } from './model/geodb-config.model';
var GeoDbProModule = /** @class */ (function () {
    function GeoDbProModule(config) {
    }
    /**
     * @param {?} config
     * @return {?}
     */
    GeoDbProModule.forRoot = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        return {
            ngModule: GeoDbProModule,
            providers: [
                { provide: GeoDbConfig, useValue: config }
            ]
        };
    };
    GeoDbProModule.decorators = [
        { type: NgModule, args: [{
                    providers: [
                        GeoDbService,
                        {
                            provide: HTTP_INTERCEPTORS,
                            useClass: AuthInterceptor,
                            multi: true
                        }
                    ],
                    imports: [
                        HttpClientModule
                    ]
                },] }
    ];
    /** @nocollapse */
    GeoDbProModule.ctorParameters = function () { return [
        { type: GeoDbConfig }
    ]; };
    return GeoDbProModule;
}());
export { GeoDbProModule };
