// wft-geodb-angular-client/esm2015/lib/geodb-pro.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClientModule, HTTP_INTERCEPTORS } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { AuthInterceptor } from './auth.interceptor';
import { GeoDbService } from './geodb.service';
import { GeoDbConfig } from './model/geodb-config.model';
export class GeoDbProModule {
    /**
     * @param {?} config
     */
    constructor(config) {
    }
    /**
     * @param {?} config
     * @return {?}
     */
    static forRoot(config) {
        return {
            ngModule: GeoDbProModule,
            providers: [
                { provide: GeoDbConfig, useValue: config }
            ]
        };
    }
}
GeoDbProModule.decorators = [
    { type: NgModule, args: [{
                providers: [
                    GeoDbService,
                    {
                        provide: HTTP_INTERCEPTORS,
                        useClass: AuthInterceptor,
                        multi: true
                    }
                ],
                imports: [
                    HttpClientModule
                ]
            },] }
];
/** @nocollapse */
GeoDbProModule.ctorParameters = () => [
    { type: GeoDbConfig }
];
